fdbserver/BlobGranuleServerCommon.actor.cpp

```diff
@@ -458,40 +458,77 @@ ACTOR Future<Void> loadBlobMetadataForTenants(BGTenantMap* self, std::vector<Blo
-		} else {
-			requestFuture = Never();
-		}
-		choose {
-			when(EKPGetLatestBlobMetadataReply rep = wait(requestFuture)) {
-				ASSERT(rep.blobMetadataDetails.size() == req.domainIds.size());
-				// not guaranteed to be in same order in the request as the response
-				for (auto& metadata : rep.blobMetadataDetails) {
-					auto info = self->tenantInfoById.find(metadata.domainId);
-					if (info == self->tenantInfoById.end()) {
-						continue;
+		try {
+			Future<EKPGetLatestBlobMetadataReply> requestFuture;
+			if (self->dbInfo.isValid() && self->dbInfo->get().encryptKeyProxy.present()) {
+				req.reply.reset();
+				requestFuture =
+				    brokenPromiseToNever(self->dbInfo->get().encryptKeyProxy.get().getLatestBlobMetadata.getReply(req));
+			} else {
+				requestFuture = Never();
+			}
+			choose {
+				when(EKPGetLatestBlobMetadataReply rep = wait(requestFuture)) {
+					ASSERT(rep.blobMetadataDetails.size() <= req.domainIds.size());
+					// not guaranteed to be in same order in the request as the response
+					for (auto& metadata : rep.blobMetadataDetails) {
+						auto info = self->tenantInfoById.find(metadata.domainId);
+						if (info == self->tenantInfoById.end()) {
+							continue;
+						}
+						auto dataEntry = self->tenantData.rangeContaining(info->second.prefix);
+						ASSERT(dataEntry.begin() == info->second.prefix);
+						dataEntry.cvalue()->updateBStore(metadata);
+					}
+					double elapsed = now() - startTime;
+					BlobCipherMetrics::getInstance()->getBlobMetadataLatency.addMeasurement(elapsed);
+
+					if (rep.blobMetadataDetails.size() == req.domainIds.size()) {
+						return Void();
+					}
+
+					// if not all tenants included in response, or on error, retry with missing ones
+					std::unordered_set<BlobMetadataDomainId> missingIds;
+					for (auto& id : req.domainIds) {
+						missingIds.insert(id);
+					}
+					for (auto& metadata : rep.blobMetadataDetails) {
+						missingIds.erase(metadata.domainId);
+					}
+					ASSERT(missingIds.size() > 0);
+
+					TraceEvent(SevWarn, "BlobMetadataFetchMissingTenants")
+					    .suppressFor(30.0)
+					    .detail("Count", missingIds.size());
+					CODE_PROBE(true, "blob metadata fetch missing tenants");
+
+					req.domainIds.clear();
+					for (auto& id : missingIds) {
+						req.domainIds.push_back(id);
 					}
-					auto dataEntry = self->tenantData.rangeContaining(info->second.prefix);
-					ASSERT(dataEntry.begin() == info->second.prefix);
-					dataEntry.cvalue()->updateBStore(metadata);
 				}
-				double elapsed = now() - startTime;
-				BlobCipherMetrics::getInstance()->getBlobMetadataLatency.addMeasurement(elapsed);
-				return Void();
+				when(wait(self->dbInfo->onChange())) {
+					// reset retry sleep
+					retrySleep = 0.1;
+				}
+			}
+		} catch (Error& e) {
+			if (e.code() == error_code_operation_cancelled) {
+				throw e;
 			}
-			when(wait(self->dbInfo->onChange())) {}
+			CODE_PROBE(true, "blob metadata fetch error");
```

Comment:
I feel we should have more fine grained control on the error type here if possible, for instance: failing due to authentication/validation errors from external KMS could mean an attacker trying to masquerade an FDB process, in such cases we should fail the whole process, for timeout/io_error/wrong credentials, I agree failing the whole process might not work.